src/sync/mpsc.rs

```diff
@@ -0,0 +1,175 @@
+        match self.tx.send(t) {
+            Ok(_) => {
+                let _ = self.inner.num_msgs_pending.fetch_add(1, SeqCst);
+                // if the rx_task has been set by the receiver, attempt to unpark the receiver
+                if !self.inner.is_first_rx_poll.load(SeqCst) {
+                    // try to get the lock. If it fails, some other sender is holding the lock
+                    // and will unpark the receiver.
+                    if let Some(mut opt_task) = self.inner.rx_task.try_lock() {
+                        // if we got the lock, then the rx_task is ours to take and it must exist
+                        // because either the receiver has set it (first time) or a prior send has
+                        // stored a cloned task
+                        let task = opt_task.take().unwrap();
+                        // clone and store the rx_task so a subsequent send will be able to wake up
+                        // the receiver too
+                        *opt_task = Some(task.clone());
+                        task.unpark();
+                    }
+                }
+                Ok(())
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        // Each cloned sender calls `Drop` so we need to ensure that only the last sender being
+        // dropped will unpark the `Receiver`.
+        if self.inner.num_senders.fetch_sub(1, SeqCst) == 1 &&
+           !self.inner.is_first_rx_poll.load(SeqCst) {
+            // wake up receiver to go to disconnection event or get more data if there is any
+            // can unwrap because we are guaranteed the receiver has set this value by now
+            let task = self.inner.rx_task_tx_drop.try_lock().unwrap().take().unwrap();
+            task.unpark();
+        }
+    }
+}
+
+/// Creates a new unbounded multiple-producer, single-consumer channel. `Sender` can be cloned.
+///
+/// This channel does not implement back pressure.
+pub fn unbounded<T>() -> (Sender<T>, Receiver<T>) {
+    let (tx, rx) = mpsc::channel();
+    let inner = Arc::new(Inner {
+        rx_task: Lock::new(None),
+        rx_task_tx_drop: Lock::new(None),
+        num_senders: AtomicUsize::new(1),
+        num_msgs_pending: AtomicUsize::new(0),
+        is_first_rx_poll: AtomicBool::new(true),
+    });
+    let tx = Sender {
+        tx: tx,
+        inner: inner.clone(),
+    };
+    let rx = Receiver {
+        rx: rx,
+        inner: inner,
+    };
+    (tx, rx)
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+    type Error = ();
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        // if this is the receiver's first poll call, store two receiver tasks:
+        // rx_task_tx_drop - used only once to wake up receiver when last Sender drops
+        // rx_task - used by senders to wake up the receiver after a message has been sent
+        if self.inner.is_first_rx_poll.load(SeqCst) {
```

Comment:
Ah I did not know that was required. I guess then I will need to add some synchronization between the Sender and Receiver for the `rx_task` then to satisfy that.